File: app/src/main/java/com/bykea/pk/partner/ui/activities/JobCallActivity.java
package com.bykea.pk.partner.ui.activities;

import android.content.Intent;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.os.SystemClock;
import android.view.View;
import android.widget.RelativeLayout;
import android.widget.Toast;

import androidx.appcompat.widget.AppCompatImageView;
import androidx.core.content.ContextCompat;

import com.bykea.pk.partner.BuildConfig;
import com.bykea.pk.partner.DriverApp;
import com.bykea.pk.partner.R;
import com.bykea.pk.partner.dal.Stop;
import com.bykea.pk.partner.dal.source.JobsDataSource;
import com.bykea.pk.partner.dal.source.JobsRepository;
import com.bykea.pk.partner.dal.source.socket.payload.JobCall;
import com.bykea.pk.partner.dal.util.Injection;
import com.bykea.pk.partner.models.response.AcceptCallResponse;
import com.bykea.pk.partner.models.response.FreeDriverResponse;
import com.bykea.pk.partner.models.response.RejectCallResponse;
import com.bykea.pk.partner.repositories.UserDataHandler;
import com.bykea.pk.partner.repositories.UserRepository;
import com.bykea.pk.partner.ui.helpers.ActivityStackManager;
import com.bykea.pk.partner.ui.helpers.AppPreferences;
import com.bykea.pk.partner.utils.Constants;
import com.bykea.pk.partner.utils.Dialogs;
import com.bykea.pk.partner.utils.Keys;
import com.bykea.pk.partner.utils.TripStatus;
import com.bykea.pk.partner.utils.Utils;
import com.bykea.pk.partner.widgets.AutoFitFontTextView;
import com.bykea.pk.partner.widgets.DonutProgress;
import com.bykea.pk.partner.widgets.FontTextView;
import com.google.gson.Gson;

import org.greenrobot.eventbus.Subscribe;
import org.json.JSONException;
import org.json.JSONObject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

import static com.bykea.pk.partner.utils.Constants.ServiceType.RIDE_CODE;
import static com.bykea.pk.partner.utils.Constants.ServiceType.SEND_CODE;
import static com.bykea.pk.partner.utils.Constants.ServiceType.SEND_COD_CODE;


public class JobCallActivity extends BaseActivity {

    public static String KEY_CALL_DATA = "KEY_CALL_DATA";
    public static String KEY_IS_FROM_PUSH = "isGcm";
    public String TAG = JobCallActivity.class.getSimpleName();
    @BindView(R.id.counterTv)
    FontTextView counterTv;
    @BindView(R.id.donut_progress)
    DonutProgress donutProgress;
    @BindView(R.id.activity_calling)
    RelativeLayout activity_calling;
    @BindView(R.id.ivCallType)
    AppCompatImageView ivCallType;
    @BindView(R.id.estArrivalTimeTV)
    FontTextView estArrivalTimeTV;
    @BindView(R.id.estDistanceTV)
    FontTextView estDistanceTV;
    @BindView(R.id.dropZoneNameTV)
    AutoFitFontTextView dropZoneNameTV;
    @BindView(R.id.acceptCallBtn)
    RelativeLayout acceptCallBtn;
    private MediaPlayer _mpSound;
    private String tripId;
    private boolean isCancelledByPassenger;
    private long mLastClickTime;
    private int RIDE_ACCEPTANCE_TIMEOUT = 20;
    int timerDuration = RIDE_ACCEPTANCE_TIMEOUT;
    private String acceptSeconds = "0";
    private JobCall jobCall;
    private UserRepository userRepo;

    private CountDownTimer timer;
    /**
     * Response handler for rest calls
     */
    private UserDataHandler handler = new UserDataHandler() {

        @Override
        public void onAck(final String msg) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    Utils.appToastDebug(JobCallActivity.this, msg);
                }
            });
        }

        @Override
        public void onFreeDriver(FreeDriverResponse freeDriverResponse) {
            JobCallActivity.this.runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    ActivityStackManager.getInstance().startHomeActivity(true, JobCallActivity.this);
                    stopSound();
                    finishActivity();
                }
            });

        }

        @Override
        public void onAcceptCall(final AcceptCallResponse acceptCallResponse) {
            onAcceptSuccess(acceptCallResponse.isSuccess(), acceptCallResponse.getMessage());
        }

        @Override
        public void onRejectCall(final RejectCallResponse rejectCallResponse) {
            onAcceptFailed(rejectCallResponse.getMessage());
        }

        @Override
        public void onError(int errorCode, final String errorMessage) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    Dialogs.INSTANCE.dismissDialog();
                    Dialogs.INSTANCE.showToast(errorMessage);
                    ActivityStackManager.getInstance().startHomeActivity(true, JobCallActivity.this);
                    stopSound();
                    finishActivity();
                }
            });
        }
    };

    @Override
    protected void onResume() {
        super.onResume();
        AppPreferences.setCallingActivityOnForeground(true);
    }

    @Override
    protected void onStop() {
        super.onStop();
        AppPreferences.setCallingActivityOnForeground(false);
    }

    @Override
    protected void onDestroy() {
        stopSound();
        if (AppPreferences.isOnTrip()) {
            AppPreferences.setIncomingCall(false);
        } else {
            AppPreferences.setIncomingCall(true);
        }
        AppPreferences.setCallingActivityOnForeground(false);
        Utils.unbindDrawables(activity_calling);
        super.onDestroy();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_calling);
        overridePendingTransition(R.anim.fade_in, R.anim.fade_out);
        ButterKnife.bind(this);

        if (getIntent() != null) {
            jobCall = (JobCall) getIntent().getSerializableExtra(KEY_CALL_DATA);
            timerDuration = jobCall.getTimer();

            if (getIntent().getBooleanExtra("isGcm", false)) {
                Utils.redLog("FCM", "Calling Activity");
                DriverApp.getApplication().connect();
                DriverApp.startLocationService(this);
            }
        }

        userRepo = new UserRepository();
        Utils.unlockScreen(this);
        AppPreferences.setStatsApiCallRequired(true);
        //To inactive driver during passenger calling state
        AppPreferences.setTripStatus(TripStatus.ON_IN_PROGRESS);

        if (Utils.isConnected(JobCallActivity.this, false))
            userRepo.requestLocationUpdate(this, handler, AppPreferences.getLatitude(), AppPreferences.getLongitude());


        setInitialData();
        startTimer();
    }

    @OnClick({R.id.acceptCallBtn})
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.acceptCallBtn:
                if (mLastClickTime != 0 && (SystemClock.elapsedRealtime() - mLastClickTime < 1000)) {
                    return;
                }
                mLastClickTime = SystemClock.elapsedRealtime();
                stopSound();
                Dialogs.INSTANCE.showLoader(JobCallActivity.this);
                acceptSeconds = counterTv.getText().toString();
                acceptJob();
                timer.cancel();
                break;
        }
    }

    @Subscribe
    public void onEvent(final Intent intent) {
        isCancelledByPassenger = true;
        JobCallActivity.this.runOnUiThread(() -> {
            if (null != intent && null != intent.getExtras()) {
                if (intent.getStringExtra("action").equalsIgnoreCase(Keys.BROADCAST_CANCEL_RIDE)
                        || intent.getStringExtra("action").equalsIgnoreCase(Keys.BROADCAST_CANCEL_BY_ADMIN)) {
                    onJobCallCancelled();
                }
            }
        });

    }

    /**
     * Finish current screen
     */
    private void finishActivity() {
        if (Utils.isConnected(JobCallActivity.this, false))
            userRepo.requestLocationUpdate(this, handler, AppPreferences.getLatitude(), AppPreferences.getLongitude());
        finish();
    }

    /**
     * Log call data to mix panel
     *
     * @param callData   Job call data
     * @param isOnAccept is post accepted
     */
    private void logMixPanelEvent(JobCall callData, boolean isOnAccept) {
        try {

            JSONObject data = new JSONObject();
            data.put("PassengerID", callData.getCustomer_id());
            data.put("DriverID", AppPreferences.getPilotData().getId());
            data.put("TripID", callData.getTrip_id());
            data.put("TripNo", callData.getBooking_no());
            data.put("PickUpLocation", callData.getPickup().getLat() + "," + callData.getPickup().getLng());
            data.put("timestamp", Utils.getIsoDate());
            if (callData.getDropoff() != null) {
                data.put("DropOffLocation", callData.getDropoff().getLat() + "," + callData.getDropoff().getLng());
            }
            data.put("ETA", "" + callData.getPickup().getDuration());
            data.put("EstimatedDistance", AppPreferences.getEstimatedDistance());
            data.put("CurrentLocation", Utils.getCurrentLocation());
            data.put("DriverName", AppPreferences.getPilotData().getFullName());
            data.put("SignUpCity", AppPreferences.getPilotData().getCity().getName());
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    /**
     * Start timer with dynamic timing, also play the calling sound
     */
    private void startTimer() {

        _mpSound = MediaPlayer.create(this, R.raw.ringtone);
        _mpSound.start();

        donutProgress.setMax(timerDuration);
        donutProgress.setProgress(0);

        timer = new CountDownTimer(timerDuration * 1000, 1000) {
            int progress = 0;

            @Override
            public void onTick(long millisUntilFinished) {
                progress++;
                if (progress < timerDuration) {
                    donutProgress.setProgress(progress);
                    counterTv.setText(String.valueOf(timerDuration - progress));
                } else {
                    timer.onFinish();
                }
            }

            @Override
            public void onFinish() {
                donutProgress.setProgress(0);
                counterTv.setText("0");
                acceptCallBtn.setEnabled(false);
                stopSound();
                Utils.setCallIncomingStateWithoutRestartingService();
                ActivityStackManager.getInstance().startHomeActivity(true, JobCallActivity.this);
                finishActivity();
            }
        };
        timer.start();
    }

    /**
     * Stop calling sound
     */
    private void stopSound() {
        if (null != _mpSound && _mpSound.isPlaying()) {
            _mpSound.stop();
        }
        if (null != timer) timer.cancel();
    }

    /**
     * Render job calling data
     */
    private void setInitialData() {
        tripId = jobCall.getTrip_id();
        Utils.redLog(TAG, "Call Data: " + new Gson().toJson(jobCall));
        logMixPanelEvent(jobCall, false);
        counterTv.setText(String.valueOf(timerDuration));

        ivCallType.setImageDrawable(ContextCompat.getDrawable(this, getJobImage(jobCall.getService_code())));
        estArrivalTimeTV.setText(getArrivalTime(jobCall.getPickup()));
        dropZoneNameTV.setText(getDropOffZoneName(jobCall.getDropoff()));
        estDistanceTV.setText(getDropOffDistance(jobCall.getDropoff()));
    }

    /**
     * Get job service image
     *
     * @param service_code Job service code
     * @return Drawable ID
     */
    private int getJobImage(int service_code) {
        if (service_code == RIDE_CODE)
            return R.drawable.ride;
        else if (service_code == SEND_CODE || service_code == SEND_COD_CODE)
            return R.drawable.bhejdo_no_caption;
        else
            return R.drawable.ride;
    }

    /**
     * Get arrival time from current location to pickup
     *
     * @param pickup Pickup stop
     * @return Displayable time in minutes
     */
    private String getArrivalTime(Stop pickup) {
        if (pickup != null && pickup.getDuration() > 0)
            return String.valueOf(pickup.getDuration() / 60);
        else
            return "1";
    }

    /**
     * Get the drop-off zone name to show
     *
     * @param dropoff Drop-off stop
     * @return Displayable zone name
     */
    private String getDropOffZoneName(Stop dropoff) {
        if (dropoff != null) {
            if (dropoff.getZone_ur() != null && !dropoff.getZone_ur().isEmpty()) {
                return dropoff.getZone_ur();
            } else if (dropoff.getZone_en() != null && !dropoff.getZone_en().isEmpty())
                return dropoff.getZone_en();
            else if (dropoff.getAddress() != null && !dropoff.getAddress().isEmpty())
                return dropoff.getAddress();
            else
                return getString(R.string.customer_btayega);
        } else
            return getString(R.string.customer_btayega);
    }

    /**
     * Get the drop-off distance from pickup stop
     *
     * @param dropoff Drop-off stop
     * @return Displayable distance in kilometer
     */
    private String getDropOffDistance(Stop dropoff) {
        if (dropoff != null && dropoff.getDistance() != 0)
            return String.valueOf(dropoff.getDistance() / 1000);
        else return "?";
    }

    /**
     * Inform server to accept the job request
     */
    private void acceptJob() {
        JobsRepository jobsRepo = Injection.INSTANCE.provideJobsRepository(getApplication().getApplicationContext());
        jobsRepo.acceptJob(tripId, Integer.valueOf(acceptSeconds), new JobsDataSource.AcceptJobCallback() {
            @Override
            public void onJobAccepted() {
                if (!isCancelledByPassenger) {
                    onAcceptSuccess(true, "Job Accepted");
                } else {
                    onJobCallCancelled();
                }
                if (BuildConfig.DEBUG)
                    Toast.makeText(getApplication().getApplicationContext(), "Job Accepted", Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onJobAcceptFailed() {
                onAcceptFailed("Job Accept Failed");
                if (BuildConfig.DEBUG)
                    Toast.makeText(getApplication().getApplicationContext(), "Job Accept Failed", Toast.LENGTH_SHORT).show();
            }
        });
    }

    /**
     * On success of job call accept
     *
     * @param success Success status
     * @param message Success message
     */
    private void onAcceptSuccess(Boolean success, String message) {
        runOnUiThread(() -> {
            Dialogs.INSTANCE.dismissDialog();
            Dialogs.INSTANCE.showTempToast(message);
            if (!isCancelledByPassenger) {
                if (success) {
                    AppPreferences.clearTripDistanceData();
                    AppPreferences.setTripStatus(TripStatus.ON_ACCEPT_CALL);
                    AppPreferences.setTripAcceptTime(System.currentTimeMillis());
                    logMixPanelEvent(jobCall, true);
                    AppPreferences.addLocCoordinateInTrip(AppPreferences.getLatitude(), AppPreferences.getLongitude());
                    AppPreferences.setIsOnTrip(true);
                    AppPreferences.setDeliveryType(Constants.CallType.SINGLE);
                    ActivityStackManager.getInstance().startJobActivity(JobCallActivity.this);
                    stopSound();
                    finishActivity();
                } else {
                    Utils.setCallIncomingState();
                    Dialogs.INSTANCE.showToast(message);
                }
            } else {
                onJobCallCancelled();
            }
        });

    }

    /**
     * On failure of accept call
     *
     * @param message Failure message
     */
    private void onAcceptFailed(String message) {
        runOnUiThread(() -> {
            Dialogs.INSTANCE.showToast(message);
            Dialogs.INSTANCE.dismissDialog();
            if (AppPreferences.isOnTrip()) {
                AppPreferences.setIncomingCall(false);
                AppPreferences.setTripStatus(TripStatus.ON_FREE);
            } else {
                AppPreferences.setIncomingCall(true);
            }
            ActivityStackManager.getInstance().startHomeActivity(true, JobCallActivity.this);
            stopSound();
            finishActivity();
        });
    }

    /**
     * On job call cancelled during job calling
     */
    private void onJobCallCancelled() {
        Utils.setCallIncomingState();
        AppPreferences.setTripStatus(TripStatus.ON_FREE);
        stopSound();
        ActivityStackManager.getInstance().startHomeActivityFromCancelTrip(false, this);
        finishActivity();
    }
}

File: app/src/main/java/com/bykea/pk/partner/ui/activities/JobCallActivity.kt
package com.bykea.pk.partner.ui.activities

import android.content.Intent
import android.media.MediaPlayer
import android.os.Bundle
import android.os.CountDownTimer
import android.os.SystemClock
import android.view.View
import android.widget.Toast
import androidx.core.content.ContextCompat
import butterknife.ButterKnife
import butterknife.OnClick
import com.bykea.pk.partner.BuildConfig
import com.bykea.pk.partner.DriverApp
import com.bykea.pk.partner.R
import com.bykea.pk.partner.dal.Stop
import com.bykea.pk.partner.dal.source.JobsDataSource
import com.bykea.pk.partner.dal.source.socket.payload.JobCall
import com.bykea.pk.partner.dal.util.Injection
import com.bykea.pk.partner.models.response.AcceptCallResponse
import com.bykea.pk.partner.models.response.FreeDriverResponse
import com.bykea.pk.partner.models.response.RejectCallResponse
import com.bykea.pk.partner.repositories.UserDataHandler
import com.bykea.pk.partner.repositories.UserRepository
import com.bykea.pk.partner.ui.helpers.ActivityStackManager
import com.bykea.pk.partner.ui.helpers.AppPreferences
import com.bykea.pk.partner.utils.*
import com.bykea.pk.partner.utils.Constants.ServiceType.*
import com.google.gson.Gson
import kotlinx.android.synthetic.main.activity_job_call.*
import org.greenrobot.eventbus.Subscribe
import org.json.JSONException
import org.json.JSONObject


class JobCallActivity : BaseActivity() {

    private var _mpSound: MediaPlayer? = null
    private var tripId: String? = null
    private var isCancelledByPassenger: Boolean = false
    private var mLastClickTime: Long = 0
    private val RIDE_ACCEPTANCE_TIMEOUT = 20
    internal var timerDuration = RIDE_ACCEPTANCE_TIMEOUT
    private var acceptSeconds = "0"
    private var jobCall: JobCall? = null
    private var userRepo: UserRepository? = null
    private var timer: CountDownTimer? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_job_call)
        overridePendingTransition(R.anim.fade_in, R.anim.fade_out)
        ButterKnife.bind(this)

        if (intent != null) {
            jobCall = intent.getSerializableExtra(KEY_CALL_DATA) as JobCall
            timerDuration = jobCall!!.timer

            if (intent.getBooleanExtra(KEY_IS_FROM_PUSH, false)) {
                DriverApp.getApplication().connect()
                DriverApp.startLocationService(this)
            }
        }

        userRepo = UserRepository()
        Utils.unlockScreen(this)
        AppPreferences.setStatsApiCallRequired(true)
        //To inactive driver during passenger calling state
        AppPreferences.setTripStatus(TripStatus.ON_IN_PROGRESS)

        if (Utils.isConnected(this@JobCallActivity, false))
            userRepo!!.requestLocationUpdate(this, handler, AppPreferences.getLatitude(), AppPreferences.getLongitude())


        setInitialData()
        startTimer()
    }

    override fun onResume() {
        super.onResume()
        AppPreferences.setCallingActivityOnForeground(true)
    }

    override fun onStop() {
        super.onStop()
        AppPreferences.setCallingActivityOnForeground(false)
    }

    override fun onDestroy() {
        stopSound()
        if (AppPreferences.isOnTrip()) {
            AppPreferences.setIncomingCall(false)
        } else {
            AppPreferences.setIncomingCall(true)
        }
        AppPreferences.setCallingActivityOnForeground(false)
        Utils.unbindDrawables(activity_calling)
        super.onDestroy()
    }

    @OnClick(R.id.acceptCallBtn)
    fun onClick(view: View) {
        when (view.id) {
            R.id.acceptCallBtn -> {
                if (mLastClickTime != 0L && SystemClock.elapsedRealtime() - mLastClickTime < 1000) {
                    return
                }
                mLastClickTime = SystemClock.elapsedRealtime()
                stopSound()
                Dialogs.INSTANCE.showLoader(this@JobCallActivity)
                acceptSeconds = counterTv!!.text.toString()
                acceptJob()
                timer!!.cancel()
            }
        }
    }

    @Subscribe
    fun onEvent(intent: Intent?) {
        isCancelledByPassenger = true
        this@JobCallActivity.runOnUiThread {
            if (null != intent && null != intent.extras) {
                if (intent.getStringExtra("action").equals(Keys.BROADCAST_CANCEL_RIDE, ignoreCase = true) || intent.getStringExtra("action").equals(Keys.BROADCAST_CANCEL_BY_ADMIN, ignoreCase = true)) {
                    onJobCallCancelled()
                }
            }
        }

    }

    /**
     * Finish current screen
     */
    private fun finishActivity() {
        if (Utils.isConnected(this@JobCallActivity, false))
            userRepo!!.requestLocationUpdate(this, handler, AppPreferences.getLatitude(), AppPreferences.getLongitude())
        finish()
    }

    /**
     * Log call data to mix panel
     *
     * @param callData   Job call data
     * @param isOnAccept is post accepted
     */
    private fun logMixPanelEvent(callData: JobCall, isOnAccept: Boolean) {
        try {

            val data = JSONObject()
            data.put("PassengerID", callData.customer_id)
            data.put("DriverID", AppPreferences.getPilotData().id)
            data.put("TripID", callData.trip_id)
            data.put("TripNo", callData.booking_no)
            data.put("PickUpLocation", callData.pickup.lat.toString() + "," + callData.pickup.lng)
            data.put("timestamp", Utils.getIsoDate())
            if (callData.dropoff != null) {
                data.put("DropOffLocation", callData.dropoff!!.lat.toString() + "," + callData.dropoff!!.lng)
            }
            data.put("ETA", "" + callData.pickup.duration)
            data.put("EstimatedDistance", AppPreferences.getEstimatedDistance())
            data.put("CurrentLocation", Utils.getCurrentLocation())
            data.put("DriverName", AppPreferences.getPilotData().fullName)
            data.put("SignUpCity", AppPreferences.getPilotData().city.name)
        } catch (e: JSONException) {
            e.printStackTrace()
        }

    }

    /**
     * Start timer with dynamic timing, also play the calling sound
     */
    private fun startTimer() {

        _mpSound = MediaPlayer.create(this, R.raw.ringtone)
        _mpSound!!.start()

        donut_progress!!.max = timerDuration
        donut_progress!!.progress = 0f

        timer = object : CountDownTimer((timerDuration * 1000).toLong(), 1000) {
            internal var progress = 0

            override fun onTick(millisUntilFinished: Long) {
                progress++
                if (progress < timerDuration) {
                    donut_progress!!.progress = progress.toFloat()
                    counterTv!!.text = (timerDuration - progress).toString()
                } else {
                    timer!!.onFinish()
                }
            }

            override fun onFinish() {
                donut_progress!!.progress = 0f
                counterTv!!.text = "0"
                acceptCallBtn!!.isEnabled = false
                stopSound()
                Utils.setCallIncomingStateWithoutRestartingService()
                ActivityStackManager.getInstance().startHomeActivity(true, this@JobCallActivity)
                finishActivity()
            }
        }
        timer!!.start()
    }

    /**
     * Stop calling sound
     */
    private fun stopSound() {
        if (null != _mpSound && _mpSound!!.isPlaying) {
            _mpSound!!.stop()
        }
        if (null != timer) timer!!.cancel()
    }

    /**
     * Render job calling data
     */
    private fun setInitialData() {
        tripId = jobCall!!.trip_id
        Utils.redLog(TAG, "Call Data: " + Gson().toJson(jobCall))
        logMixPanelEvent(jobCall!!, false)
        counterTv!!.text = timerDuration.toString()

        ivCallType!!.setImageDrawable(ContextCompat.getDrawable(this, getJobImage(jobCall!!.service_code)))
        estArrivalTimeTV!!.text = getArrivalTime(jobCall!!.pickup)
        dropZoneNameTV!!.text = getDropOffZoneName(jobCall!!.dropoff) + ""
        estDistanceTV!!.text = getDropOffDistance(jobCall!!.dropoff)
    }

    /**
     * Get job service image
     *
     * @param service_code Job service code
     * @return Drawable ID
     */
    private fun getJobImage(service_code: Int): Int {
        return if (service_code == RIDE_CODE)
            R.drawable.ride
        else if (service_code == SEND_CODE || service_code == SEND_COD_CODE)
            R.drawable.bhejdo_no_caption
        else
            R.drawable.ride
    }

    /**
     * Get arrival time from current location to pickup
     *
     * @param pickup Pickup stop
     * @return Displayable time in minutes
     */
    private fun getArrivalTime(pickup: Stop?): String {
        return if (pickup != null && pickup.duration > 0)
            (pickup.duration / 60).toString()
        else
            "1"
    }

    /**
     * Get the drop-off zone name to show
     *
     * @param dropoff Drop-off stop
     * @return Displayable zone name
     */
    private fun getDropOffZoneName(dropoff: Stop?): String? {
        return if (dropoff != null) {
            if (dropoff.zone_ur != null && !dropoff.zone_ur!!.isEmpty()) {
                dropoff.zone_ur
            } else if (dropoff.zone_en != null && !dropoff.zone_en!!.isEmpty())
                dropoff.zone_en
            else if (dropoff.address != null && !dropoff.address!!.isEmpty())
                dropoff.address
            else
                getString(R.string.customer_btayega)
        } else
            getString(R.string.customer_btayega)
    }

    /**
     * Get the drop-off distance from pickup stop
     *
     * @param dropoff Drop-off stop
     * @return Displayable distance in kilometer
     */
    private fun getDropOffDistance(dropoff: Stop?): String {
        return if (dropoff != null && dropoff.distance != 0)
            (dropoff.distance / 1000).toString()
        else
            "?"
    }

    /**
     * Inform server to accept the job request
     */
    private fun acceptJob() {
        val jobsRepo = Injection.provideJobsRepository(application.applicationContext)
        jobsRepo.acceptJob(tripId!!, Integer.valueOf(acceptSeconds), object : JobsDataSource.AcceptJobCallback {
            override fun onJobAccepted() {
                if (!isCancelledByPassenger) {
                    onAcceptSuccess(true, "Job Accepted")
                } else {
                    onJobCallCancelled()
                }
                if (BuildConfig.DEBUG)
                    Toast.makeText(application.applicationContext, "Job Accepted", Toast.LENGTH_SHORT).show()
            }

            override fun onJobAcceptFailed() {
                onAcceptFailed("Job Accept Failed")
                if (BuildConfig.DEBUG)
                    Toast.makeText(application.applicationContext, "Job Accept Failed", Toast.LENGTH_SHORT).show()
            }
        })
    }

    /**
     * On success of job call accept
     *
     * @param success Success status
     * @param message Success message
     */
    private fun onAcceptSuccess(success: Boolean?, message: String) {
        runOnUiThread {
            Dialogs.INSTANCE.dismissDialog()
            Dialogs.INSTANCE.showTempToast(message)
            if (!isCancelledByPassenger) {
                if (success!!) {
                    AppPreferences.clearTripDistanceData()
                    AppPreferences.setTripStatus(TripStatus.ON_ACCEPT_CALL)
                    AppPreferences.setTripAcceptTime(System.currentTimeMillis())
                    logMixPanelEvent(jobCall!!, true)
                    AppPreferences.addLocCoordinateInTrip(AppPreferences.getLatitude(), AppPreferences.getLongitude())
                    AppPreferences.setIsOnTrip(true)
                    AppPreferences.setDeliveryType(Constants.CallType.SINGLE)
                    ActivityStackManager.getInstance().startJobActivity(this@JobCallActivity)
                    stopSound()
                    finishActivity()
                } else {
                    Utils.setCallIncomingState()
                    Dialogs.INSTANCE.showToast(message)
                }
            } else {
                onJobCallCancelled()
            }
        }

    }

    /**
     * On failure of accept call
     *
     * @param message Failure message
     */
    private fun onAcceptFailed(message: String) {
        runOnUiThread {
            Dialogs.INSTANCE.showToast(message)
            Dialogs.INSTANCE.dismissDialog()
            if (AppPreferences.isOnTrip()) {
                AppPreferences.setIncomingCall(false)
                AppPreferences.setTripStatus(TripStatus.ON_FREE)
            } else {
                AppPreferences.setIncomingCall(true)
            }
            ActivityStackManager.getInstance().startHomeActivity(true, this@JobCallActivity)
            stopSound()
            finishActivity()
        }
    }

    /**
     * On job call cancelled during job calling
     */
    private fun onJobCallCancelled() {
        Utils.setCallIncomingState()
        AppPreferences.setTripStatus(TripStatus.ON_FREE)
        stopSound()
        ActivityStackManager.getInstance().startHomeActivityFromCancelTrip(false, this)
        finishActivity()
    }

    /**
     * Response handler for rest calls
     */
    private val handler = object : UserDataHandler() {

        override fun onAck(msg: String) {
            runOnUiThread { Utils.appToastDebug(this@JobCallActivity, msg) }
        }

        override fun onFreeDriver(freeDriverResponse: FreeDriverResponse) {
            this@JobCallActivity.runOnUiThread {
                ActivityStackManager.getInstance().startHomeActivity(true, this@JobCallActivity)
                stopSound()
                finishActivity()
            }

        }

        override fun onAcceptCall(acceptCallResponse: AcceptCallResponse) {
            onAcceptSuccess(acceptCallResponse.isSuccess, acceptCallResponse.message)
        }

        override fun onRejectCall(rejectCallResponse: RejectCallResponse) {
            onAcceptFailed(rejectCallResponse.message)
        }

        override fun onError(errorCode: Int, errorMessage: String) {
            runOnUiThread {
                Dialogs.INSTANCE.dismissDialog()
                Dialogs.INSTANCE.showToast(errorMessage)
                ActivityStackManager.getInstance().startHomeActivity(true, this@JobCallActivity)
                stopSound()
                finishActivity()
            }
        }
    }

    companion object {
        var TAG: String = JobCallActivity::class.java.simpleName
        var KEY_CALL_DATA = "KEY_CALL_DATA"
        var KEY_IS_FROM_PUSH = "isGcm"
    }
}
